runtime/complex.go

```diff
@@ -130,3 +134,133 @@ func complexCoerce(o *Object) (complex128, bool) {
+			result, raised := floatConvert(floatSlot, f, o)
+			if raised != nil {
+				return nil, raised
+			}
+			f := result.Value()
+			return NewComplex(complex(f, 0)).ToObject(), nil
+		}
+		if !o.isInstance(StrType) {
+			return nil, f.RaiseType(TypeErrorType, "complex() argument must be a string or a number")
+		}
+		s := toStrUnsafe(o).Value()
+		result, err := parseComplex(s)
+		if err != nil {
+			return nil, f.RaiseType(ValueErrorType, "complex() arg is a malformed string")
+		}
+		return NewComplex(result).ToObject(), nil
+	}
+
+	// TODO: fix me
+	return nil, nil
+}
+
+// ParseComplex converts the string s to a complex number.
+// If string is well-formed (one of these forms: <float>, <float>j,
+// <float><signed-float>j, <float><sign>j, <sign>j or j, where <float> is
+// any numeric string that's acceptable by strconv.ParseFloat(s, 64)),
+// ParseComplex returns the respective complex128 number.
+func parseComplex(s string) (complex128, error) {
+	c := strings.Count(s, "(")
+	if (c > 1) || (c == 1 && strings.Count(s, ")") != 1) {
+		return complex(0, 0), errors.New("Malformed complex string, more than one matching parantheses")
+	}
+	ts := strings.TrimSpace(s)
+	ts = strings.Trim(ts, "()")
+	ts = strings.TrimSpace(ts)
+	re := `(?i)(?:(?:(?:(?:\d*\.\d+)|(?:\d+\.?))(?:[Ee][+-]?\d+)?)|(?:infinity)|(?:nan)|(?:inf))`
+	fre := `[-+]?` + re
+	sre := `[-+]` + re
+	fsfj := `(?:(?P<real1>` + fre + `)(?P<imag1>` + sre + `)j)`
+	fsj := `(?:(?P<real2>` + fre + `)(?P<sign2>[-+])j)`
+	fj := `(?P<imag3>` + fre + `)j`
+	f := `(?P<real4>` + fre + `)`
+	sj := `(?P<sign5>[-+])j`
+	j := `(?P<onlyJ>j)`
+	r := regexp.MustCompile(`^(?:` + fsfj + `|` + fsj + `|` + fj + `|` + f + `|` + sj + `|` + j + `)$`)
+	subs := r.FindStringSubmatch(ts)
+	if subs == nil {
+		return complex(0, 0), errors.New("Malformed complex string, no mathing pattern found")
+	}
+	const Real1 = 1
+	const Imag1 = 2
+	const Real2 = 3
+	const Sign2 = 4
+	const Imag3 = 5
+	const Real4 = 6
+	const Sign5 = 7
+	const OnlyJ = 8
+	if subs[Real1] != "" && subs[Imag1] != "" {
+		r, _ := strconv.ParseFloat(unsignNaN(subs[Real1]), 64)
+		i, err := strconv.ParseFloat(unsignNaN(subs[Imag1]), 64)
+		return complex(r, i), err
+	}
+	if subs[Real2] != "" && subs[Sign2] != "" {
+		r, err := strconv.ParseFloat(unsignNaN(subs[Real2]), 64)
+		if subs[Sign2] == "-" {
+			return complex(r, -1), err
+		}
+		return complex(r, 1), err
+	}
+	if subs[Imag3] != "" {
+		i, err := strconv.ParseFloat(unsignNaN(subs[Imag3]), 64)
+		return complex(0, i), err
+	}
+	if subs[Real4] != "" {
+		r, err := strconv.ParseFloat(unsignNaN(subs[Real4]), 64)
+		return complex(r, 0), err
+	}
+	if subs[Sign5] != "" {
+		if subs[Sign5] == "-" {
+			return complex(0, -1), nil
+		}
+		return complex(0, 1), nil
+	}
+	if subs[OnlyJ] != "" {
+		return complex(0, 1), nil
+	}
+	return complex(0, 0), errors.New("Malformed complex string")
+}
+
+func unsignNaN(s string) string {
+	us := strings.ToUpper(s)
```

Comment:
Nitpick: In this codebase we usually use ToLower() for this kind of normalization.